pkg/storage/internal/actions/abort.go

```diff
@@ -0,0 +1,88 @@
+package actions
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/logging"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/storage/entity"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/storage/history"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/wdk"
+)
+
+type abortAction struct {
+	logger           *slog.Logger
+	transactionsRepo TransactionsRepo
+}
+
+func newAbortAction(logger *slog.Logger, transactions TransactionsRepo) *abortAction {
+	return &abortAction{
+		logger:           logging.Child(logger, "abort_action"),
+		transactionsRepo: transactions,
+	}
+}
+
+func (a *abortAction) AbortAction(ctx context.Context, userID int, args *wdk.AbortActionArgs) (*wdk.AbortActionResult, error) {
+	txEntity, err := a.transactionsRepo.FindUniqueTransactionByReference(ctx, userID, *args.Reference)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find unique transaction by reference: %w", err)
+	}
+
+	if txEntity == nil && len(*args.Reference) == 64 {
+		txEntity, err = a.transactionsRepo.FindTransactionByUserIDAndTxID(ctx, userID, *args.Reference)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find transaction by txid: %w", err)
+		}
+	}
+
+	if txEntity == nil {
+		return nil, fmt.Errorf("failed to find unique transaction by reference: expected exactly one transaction with reference %s, found 0", *args.Reference)
+	}
+
+	err = validateTxForAbort(txEntity)
+	if err != nil {
+		return nil, err
+	}
+
+	history := history.NewBuilder().AbortAction(*args.Reference)
+
+	if txEntity.TxID != nil {
+		err = a.transactionsRepo.UpdateTransactionStatusForTxID(ctx, *txEntity.TxID, wdk.TxStatusFailed, wdk.ProvenTxStatusInvalid, history)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update known transaction status: %w", err)
+		}
+	}
+
+	err = a.transactionsRepo.UpdateTransactionStatusByID(ctx, txEntity.ID, wdk.TxStatusFailed, wdk.ProvenTxStatusInvalid)
```

Comment:
maybe I miss something, but if the txEntity has TxID, then we're updating the status twice?

IMHO you should use separated method for UpdateKnownTxStatus in case TxID exists.